After 17 days of continuous operation, some topics are published with a message indicating that i/o timeout and cannot be published again

## Which operating system are you using?

<!-- fill checkboxes with a x. Example: [x] Linux -->

OS

- [x] Linux

Architecture

- [x] arm64

## Describe the issue

Description
goroslib version:v0.0.0-20220623231355-7a670441a069

Hello, I have a problem with my program, I need your help, thank you

After 17 days of continuous operation, some topics are published with a message indicating that i/o timeout and cannot be published again, Data accumulation but failure to publish results in high CPU and memory usage

`[2022/11/22 16:33:38] [ERROR] publisher '/hardware_status' is unable to write to subscriber '127.0.0.1:36040' (TCP): write tcp 127.0.0.1:39843->127.0.0.1:36040: i/o timeout

[2022/11/22 16:33:39] [ERROR] publisher '/hardware_status' is unable to write to subscriber '127.0.0.1:36040' (TCP): write tcp 127.0.0.1:39843->127.0.0.1:36040: i/o timeout

[2022/11/22 16:33:40] [ERROR] publisher '/hardware_status' is unable to write to subscriber '127.0.0.1:36040' (TCP): write tcp 127.0.0.1:39843->127.0.0.1:36040: i/o timeout

[2022/11/22 16:33:41] [ERROR] publisher '/hardware_status' is unable to write to subscriber '127.0.0.1:36040' (TCP): write tcp 127.0.0.1:39843->127.0.0.1:36040: i/o timeout

[2022/11/22 16:33:42] [ERROR] publisher '/hardware_status' is unable to write to subscriber '127.0.0.1:36040' (TCP): write tcp 127.0.0.1:39843->127.0.0.1:36040: i/o timeout

[2022/11/22 16:33:43] [ERROR] publisher '/hardware_status' is unable to write to subscriber '127.0.0.1:36040' (TCP): write tcp 127.0.0.1:39843->127.0.0.1:36040: i/o timeout

[2022/11/22 16:33:44] [ERROR] publisher '/hardware_status' is unable to write to subscriber '127.0.0.1:36040' (TCP): write tcp 127.0.0.1:39843->127.0.0.1:36040: i/o timeout

[2022/11/22 16:33:45] [ERROR] publisher '/hardware_status' is unable to write to subscriber '127.0.0.1:36040' (TCP): write tcp 127.0.0.1:39843->127.0.0.1:36040: i/o timeout

[2022/11/22 16:34:03] [ERROR] publisher '/rosout' is unable to write to subscriber '127.0.0.1:34718' (TCP): write tcp 127.0.0.1:39843->127.0.0.1:34718: i/o timeout

[2022/11/22 16:34:27] [ERROR] publisher '/hardware_status' is unable to write to subscriber '127.0.0.1:36040' (TCP): write tcp 127.0.0.1:39843->127.0.0.1:36040: i/o timeout

[2022/11/23 06:37:35] [ERROR] publisher '/alarm' is unable to write to subscriber '127.0.0.1:35988' (TCP): write tcp 127.0.0.1:39843->127.0.0.1:35988: i/o timeout`

--- a/publisher.go
+++ b/publisher.go
@@ -331,7 +331,10 @@ outer:
 				p.subscribers[req.header.Callerid] = ps
 
 				if p.conf.Latch && p.lastMessage != nil {
-					ps.writeMessage(p.lastMessage)
+					select {
+					case ps.writeMessage <- p.lastMessage:
+					case <-ps.ctx.Done():
+					}
 				}
 
 				return nil
@@ -360,8 +363,11 @@ outer:
 				p.lastMessage = msg
 			}
 
-			for _, s := range p.subscribers {
-				s.writeMessage(msg)
+			for _, ps := range p.subscribers {
+				select {
+				case ps.writeMessage <- msg:
+				case <-ps.ctx.Done():
+				}
 			}
 
 		case <-p.ctx.Done():

--- a/publisher_sub.go
+++ b/publisher_sub.go
@@ -16,9 +16,11 @@ type publisherSubscriber struct {
 	tcpTConn *prototcp.Conn
 	udpAddr  *net.UDPAddr
 
-	ctx          context.Context
-	ctxCancel    func()
-	curMessageID uint8
+	ctx       context.Context
+	ctxCancel func()
+
+	// in
+	writeMessage chan interface{}
 }
 
 func newPublisherSubscriber(
@@ -31,13 +33,14 @@ func newPublisherSubscriber(
 	ctx, ctxCancel := context.WithCancel(pub.ctx)
 
 	ps := &publisherSubscriber{
-		pub:       pub,
-		callerID:  callerID,
-		tcpNConn:  tcpNConn,
-		tcpTConn:  tcpTConn,
-		udpAddr:   udpAddr,
-		ctx:       ctx,
-		ctxCancel: ctxCancel,
+		pub:          pub,
+		callerID:     callerID,
+		tcpNConn:     tcpNConn,
+		tcpTConn:     tcpTConn,
+		udpAddr:      udpAddr,
+		ctx:          ctx,
+		ctxCancel:    ctxCancel,
+		writeMessage: make(chan interface{}),
 	}
 
 	pub.subscribersWg.Add(1)
@@ -90,6 +93,7 @@ func (ps *publisherSubscriber) runTCP() error {
 	ps.tcpNConn.SetReadDeadline(time.Time{})
 
 	readerErr := make(chan error)
+
 	go func() {
 		readerErr <- func() error {
 			buf := make([]byte, 64)
@@ -102,48 +106,67 @@ func (ps *publisherSubscriber) runTCP() error {
 		}()
 	}()
 
+	writerErr := make(chan error)
+	writerClose := make(chan struct{})
+
+	go func() {
+		writerErr <- func() error {
+			for {
+				select {
+				case msg := <-ps.writeMessage:
+					ps.tcpNConn.SetWriteDeadline(time.Now().Add(writeTimeout))
+					err := ps.tcpTConn.WriteMessage(msg)
+					if err != nil {
+						return err
+					}
+
+				case <-writerClose:
+					return fmt.Errorf("terminated")
+				}
+			}
+		}()
+	}()
+
 	select {
 	case err := <-readerErr:
 		ps.tcpNConn.Close()
+		close(writerClose)
+		<-writerErr
+		return err
+
+	case err := <-writerErr:
+		ps.tcpNConn.Close()
+		<-readerErr
 		return err
 
 	case <-ps.ctx.Done():
 		ps.tcpNConn.Close()
 		<-readerErr
+		close(writerClose)
+		<-writerErr
 		return fmt.Errorf("terminated")
 	}
 }
 
 func (ps *publisherSubscriber) runUDP() error {
-	<-ps.ctx.Done()
-	return fmt.Errorf("terminated")
-}
+	curMessageID := uint8(0)
 
-func (ps *publisherSubscriber) writeMessage(msg interface{}) {
-	if ps.tcpNConn != nil {
-		ps.tcpNConn.SetWriteDeadline(time.Now().Add(writeTimeout))
-		err := ps.tcpTConn.WriteMessage(msg)
-		if err != nil {
-			ps.pub.conf.Node.Log(LogLevelError,
-				"publisher '%s' is unable to write to subscriber %s: %s",
-				ps.pub.conf.Node.absoluteTopicName(ps.pub.conf.Topic),
-				ps.subscriberLabel(),
-				err)
-		}
-	} else {
-		ps.curMessageID++
+	for {
+		select {
+		case msg := <-ps.writeMessage:
+			curMessageID++
 
-		err := ps.pub.conf.Node.udprosConn.WriteMessage(
-			ps.pub.id,
-			ps.curMessageID,
-			msg,
-			ps.udpAddr)
-		if err != nil {
-			ps.pub.conf.Node.Log(LogLevelError,
-				"publisher '%s' is unable to write to subscriber %s: %s",
-				ps.pub.conf.Node.absoluteTopicName(ps.pub.conf.Topic),
-				ps.subscriberLabel(),
-				err)
+			err := ps.pub.conf.Node.udprosConn.WriteMessage(
+				ps.pub.id,
+				curMessageID,
+				msg,
+				ps.udpAddr)
+			if err != nil {
+				return err
+			}
+
+		case <-ps.ctx.Done():
+			return fmt.Errorf("terminated")
 		}
 	}
 }

